### PR TITLE
arraynode event integration in text editor

### PR DESCRIPTION
**Pull request by daesunp**
## Description

Integrates the array node delta event payload into the text-editor app

### CODE REVIEW

File: examples/data-objects/text-editor/src/app.tsx [hunk 1]
```diff
@@ -38,23 +38,12 @@ import { type FC, useEffect, useMemo, useState } from "react";
...
-	if (hostname.endsWith(".app.github.dev")) {
-		const match = /^(.+)-\d+\.app\.github\.dev$/.exec(hostname);
-		if (match) {
-			const codespaceName = match[1];
-			return `https://${codespaceName}-${tinyliciousPort}.app.github.dev`;
-		}
-	}
-
-	return `http://localhost:${tinyliciousPort}`;
+	return window.location.origin;
```
Comment (Josmithr) [inline examples/data-objects/text-editor/src/app.tsx]: Why are we changing this?

File: packages/dds/tree/api-report/tree.alpha.api.md [hunk 2]
```diff
@@ -1582,17 +1582,36 @@ export namespace TableSchema {
...
 // @alpha
 export namespace TextAsTree {
+    // @sealed
     export interface Members {
         characterCount(): number;
         characters(): Iterable<string>;
         charactersCopy(): string[];
         fullString(): string;
         insertAt(index: number, additionalCharacters: string): void;
+        onCharactersChanged(callback: (ops: readonly TextOp[] | undefined) => void): () => void;
```
Comment (Josmithr) [inline packages/dds/tree/api-report/tree.alpha.api.md:1602]: Nit: I think it would probably be fine to just call this `onChange`. It isn't (currently) possible to put anything except string leaf nodes in a text tree, so we don't need to worry about differentiating between deep/shallow changes here. But this is just an opinion. @CraigMacomber curious what you think.

File: packages/framework/react/src/text/plain/plainUtils.ts [hunk 3]
```diff
@@ -6,7 +6,11 @@
 import type { TextAsTree } from "@fluidframework/tree/internal";
 
 /**
- * Sync `newText` into the provided `root` tree.
+ * Sync `newText` into the provided `root` tree by applying the minimal remove + insert pair
```
Comment (Josmithr) [inline packages/framework/react/src/text/plain/plainUtils.ts:9]: Thanks for the docs updates! These are very useful!

File: packages/framework/react/src/text/plain/plainUtils.ts [hunk 4]
```diff
@@ -6,7 +6,11 @@
...
+ * needed to transform the tree's current content into `newText`.
+ * @remarks
+ * The diff is computed by finding the longest shared prefix/suffix between current and new content
```
Comment (Josmithr) [inline packages/framework/react/src/text/plain/plainUtils.ts:12]: So this doesn't attempt to break up isolated ranges of changes? It just replaces everything except leading/trailing unchanged ranges?

File: packages/framework/quill-react/src/plain/quillView.tsx [hunk 5]
```diff
@@ -27,90 +28,113 @@ export interface MainViewProps {
...
-		if (quillText !== treeTextWithNewline) {
 			isUpdatingRef.current = true;
-
-			const selection = quillRef.current.getSelection();
-			quillRef.current.setText(treeTextWithNewline);
-			if (selection) {
-				const length = quillRef.current.getLength();
-				const newPosition = Math.min(selection.index, length - 1);
-				quillRef.current.setSelection(newPosition, 0);
+			try {
```
Comment (Josmithr) [inline packages/framework/quill-react/src/plain/quillView.tsx]: I'm guessing this logic is basically the same as in `quillFormattedView` and `plainTextView`? Can we share any of this?

File: packages/dds/tree/api-report/tree.alpha.api.md (hunk 2 above)
Comment (daesunp) [reply packages/dds/tree/api-report/tree.alpha.api.md:1602]: Makes sense! Would we want to have a separate `onCharactersChanged` for rich text though? Since in those cases we would have deep changes (bold, italics, etc.)

File: packages/framework/react/src/text/plain/plainUtils.ts (hunk 3 above)
Comment (daesunp) [reply packages/framework/react/src/text/plain/plainUtils.ts:9]: No problem :)

File: packages/framework/react/src/text/plain/plainUtils.ts (hunk 4 above)
Comment (daesunp) [reply packages/framework/react/src/text/plain/plainUtils.ts:12]: Yes, I've updated the docs to be more clear!

File: packages/framework/quill-react/src/plain/quillView.tsx (hunk 5 above)
Comment (daesunp) [reply packages/framework/quill-react/src/plain/quillView.tsx]: Unfortunately the smallest shared piece I found that doesn't add useCallback overhead is the isUpdating + try/finally + early-return boilerplate (which is now updated)

File: examples/data-objects/text-editor/src/app.tsx [hunk 6]
```diff
@@ -38,23 +38,12 @@ import { type FC, useEffect, useMemo, useState } from "react";
...
 }
```
[the remark this answers is not in the document]
Comment (daesunp) [reply examples/data-objects/text-editor/src/app.tsx:59]: reverted

File: examples/data-objects/text-editor/src/app.tsx (hunk 1 above)
Comment (daesunp) [reply examples/data-objects/text-editor/src/app.tsx]: reverted

File: packages/dds/tree/src/text/textDomainFormatted.ts [hunk 7]
```diff
@@ -177,6 +186,22 @@ class TextNode
...
+		);
+	}
+
+	public onContentChanged(
+		callback: (ops: readonly TextAsTree.TextOp[] | undefined) => void,
+	): () => void {
+		return TreeAlpha.on(this.content, "treeChanged", ({ delta }) =>
+			processCharactersChangedDelta(delta, (i) => this.content[i]?.content.content, callback),
+		);
+	}
```
[the remark this answers is not in the document]
Comment (daesunp) [reply packages/dds/tree/src/text/textDomainFormatted.ts:228]: Extracted an atomChar() helper

File: packages/dds/tree/src/text/textDomain.ts [hunk 8]
```diff
@@ -186,6 +200,60 @@ class StringArray extends sf.array("StringArray", SchemaFactory.string) {
...
+				const ch = getChar(readPos);
+				if (ch === undefined) {
+					// Tree is out of sync with the delta — fall back to full re-read.
+					callback(undefined);
+					return;
+				}
+				text += ch;
+				readPos++;
+			}
+			ops.push({ type: "insert", text });
```
[the remark this answers is not in the document]
Comment (daesunp) [reply packages/dds/tree/src/text/textDomain.ts]: insert-loop now accumulates characters into an array and joins at the end

File: packages/dds/tree/src/text/codePointUtils.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Returns the number of Unicode code points in `str`.
+ * @remarks
+ * Use this to translate a JavaScript string length (which is in UTF-16 code units) into
+ * the atom/code-point space used by {@link TextAsTree}.
+ * @internal
+ */
+export function codePointCount(str: string): number {
+	return [...str].length;
```
[the remark this answers is not in the document]
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts]: updated with suggestion

File: packages/dds/tree/src/text/codePointUtils.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+ */
+export function utf16LengthForCodePoints(str: string, start: number, count: number): number {
+	assert(start >= 0 && start <= str.length, "start must be within str bounds");
+	assert(count >= 0, "count must be non-negative");
+	let utf16 = 0;
+	let counted = 0;
+	while (counted < count && start + utf16 < str.length) {
+		utf16 += (str.codePointAt(start + utf16) ?? 0) > 0xffff ? 2 : 1;
+		counted++;
+	}
```
[the remark this answers is not in the document]
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts:74]: updated with assert

File: packages/dds/tree/api-report/tree.alpha.api.md (hunk 2 above)
Comment (Josmithr) [reply packages/dds/tree/api-report/tree.alpha.api.md:1602]: Yeah, I think separate / more explicit methods names makes sense in the formatted case.

File: packages/framework/react/src/text/plain/plainUtils.ts (hunk 4 above)
Comment (Josmithr) [reply packages/framework/react/src/text/plain/plainUtils.ts:12]: Thanks!

File: packages/dds/tree/src/text/codePointUtils.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import { assert } from "@fluidframework/core-utils/internal";
+
+/**
+ * Returns the number of Unicode code points in `str`.
+ * @remarks
+ * Use this to translate a JavaScript string length (which is in UTF-16 code units) into
+ * the atom/code-point space used by {@link TextAsTree}.
+ * @internal
+ */
+export function codePointCount(str: string): number {
```
Comment (Josmithr) [inline packages/dds/tree/src/text/codePointUtils.ts]: Nit: our guidelines call to avoid abbreviations when reasonable. What do you think about `value` instead of `str` here?

Comment (Josmithr) [reply packages/dds/tree/src/text/codePointUtils.ts]: (same for the function below)

File: packages/dds/tree/src/text/codePointUtils.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+ * Use this to translate {@link TextAsTree}-space counts (code points) into JavaScript string indices (UTF-16).
+ * One code point outside the Basic Multilingual Plane (e.g. most emoji) occupies two UTF-16 code units.
+ *
+ * Asserts that the requested `count` code points are fully consumable from `start`; silent truncation
+ * would misalign delta offsets applied to strings rather than surface the drift to the caller.
+ * @param str - The string to measure.
+ * @param start - The UTF-16 index in `str` to start measuring from. Must be in `[0, str.length]`.
+ * @param count - The number of Unicode code points to measure. Must be non-negative, and there must
+ * be at least `count` code points available in `str` starting at `start`.
+ * @internal
```
Comment (Josmithr) [inline packages/dds/tree/src/text/codePointUtils.ts:56]: Nit: I think some examples (using `@example` comment blocks) would likely be useful here, especially if we're planning to eventually export this as a part of the user-facing API.

File: packages/dds/tree/src/text/codePointUtils.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ */
+export function utf16LengthForCodePoints(str: string, start: number, count: number): number {
+	assert(start >= 0 && start <= str.length, "start must be within str bounds");
+	assert(count >= 0, "count must be non-negative");
+	let utf16 = 0;
+	let counted = 0;
+	while (counted < count && start + utf16 < str.length) {
+		utf16 += (str.codePointAt(start + utf16) ?? 0) > 0xffff ? 2 : 1;
```
Comment (Josmithr) [inline packages/dds/tree/src/text/codePointUtils.ts]: Nit: I think a quick comment explaining this line would be useful.

File: packages/dds/tree/src/text/codePointUtils.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+ * Asserts that the requested `count` code points are fully consumable from `start`; silent truncation
+ * would misalign delta offsets applied to strings rather than surface the drift to the caller.
+ * @param str - The string to measure.
+ * @param start - The UTF-16 index in `str` to start measuring from. Must be in `[0, str.length]`.
+ * @param count - The number of Unicode code points to measure. Must be non-negative, and there must
+ * be at least `count` code points available in `str` starting at `start`.
+ * @internal
+ */
+export function utf16LengthForCodePoints(str: string, start: number, count: number): number {
+	assert(start >= 0 && start <= str.length, "start must be within str bounds");
```
Comment (Josmithr) [inline packages/dds/tree/src/text/codePointUtils.ts]: Is the intention to export this from the package? If so, we should update these asserts to throw `UsageErrors`, in which case we can also provide specific numbers in the error message.

File: packages/dds/tree/src/text/codePointUtils.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
+/*!
```
Comment (Josmithr) [inline packages/dds/tree/src/text/codePointUtils.ts:1]: We should add a suite of unit tests for these functions.

File: packages/dds/tree/src/index.ts [hunk 16]
```diff
@@ -403,5 +403,10 @@ export { FluidSerializableAsTree } from "./serializableDomainSchema.js";
 export { TableSchema, type System_TableSchema } from "./tableSchema.js";
 export { asAlpha, asBeta } from "./api.js";
 
-export { TextAsTree, FormattedTextAsTree } from "./text/index.js";
+export {
+	TextAsTree,
+	FormattedTextAsTree,
+	codePointCount,
+	utf16LengthForCodePoints,
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/dds/tree/src/index.ts:411]: `@internal` *is* a valid release tag. It indicates that something is only intended for use within the repo. This is correct.

File: packages/framework/quill-react/src/shared/runOnce.ts [hunk 17]
```diff
@@ -0,0 +1,29 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { MutableRefObject } from "react";
+
+/**
+ * Runs `fn` exactly when the shared `isUpdatingRef` is not already set, with the ref
```
Comment (Josmithr) [inline packages/framework/quill-react/src/shared/runOnce.ts]:
```suggestion
 * Runs `fn` exactly once when the shared `isUpdatingRef` is not already set, with the ref
```

File: packages/framework/quill-react/src/plain/quillView.tsx [hunk 18]
```diff
@@ -94,34 +91,54 @@ const TextEditorView = withMemoizedTreeObservations(({ root }: { root: TextAsTre
...
-
-			isUpdatingRef.current = false;
-		}
-	}
+	// Subscribe to incremental tree changes — sync tree → Quill.
+	// Kept in a separate effect so it re-subscribes correctly after React strict mode cleanup,
+	// independent of the Quill initialization guard above.
+	useEffect(() => {
+		return root.onCharactersChanged((ops) => {
+			runOnce(isUpdatingRef, () => {
```
Comment (Josmithr) [inline packages/framework/quill-react/src/plain/quillView.tsx]: I'm not following this. Don't we only want to run this when *not* `isUpdatingRef`? Doesn't the presence of that flag indicate that the update is happening as a result of a quill state update, rather than a tree update?

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx [hunk 19]
```diff
@@ -308,6 +316,168 @@ function formatToQuillAttributes(
...
+ */
+function formatToFullQuillAttributes(
+	format: FormattedTextAsTree.CharacterFormat,
+): Record<string, unknown> {
+	// Quill uses `null` (not `undefined`) to clear attributes, so we must use null
+	// for default-valued properties rather than omitting them.
+	// eslint-disable-next-line unicorn/no-null
+	const off = null;
+
+	const sizeValue =
```
Comment (Josmithr) [inline packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Nit: would it make sense to hoist this out into a helper function?

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx [hunk 20]
```diff
@@ -308,6 +316,168 @@ function formatToQuillAttributes(
...
 
+/**
+ * Convert a CharacterFormat to Quill attributes that fully describe the formatting state.
+ * Unlike {@link formatToQuillAttributes}, this includes `null` for default-valued properties
+ * so that Quill clears any previously-set attributes. Used when applying formatting changes
+ * to retained (already-present) content via `updateContents`.
+ */
+function formatToFullQuillAttributes(
+	format: FormattedTextAsTree.CharacterFormat,
+): Record<string, unknown> {
```
Comment (Josmithr) [inline packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Is there a more specific type we can use here? Or does Quill not have anything for this?

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx [hunk 21]
```diff
@@ -308,6 +316,168 @@ function formatToQuillAttributes(
...
+ * display matches the new tree state.
+ *
+ * `preEditContent` is the Quill document text (via `quill.getText()`) captured
+ * before applying these ops. It is needed to compute the UTF-16 width of removed
+ * atoms, which are no longer present in the tree after the edit.
+ *
+ * Returns `undefined` when the tree is out of sync with the delta (e.g., a concurrent edit raced with event delivery).
+ * The caller should fall back to a full diff in that case.
+ */
+function contentOpsToQuillDelta(
```
Comment (Josmithr) [inline packages/framework/quill-react/src/formatted/quillFormattedView.tsx:128]: Nit: this seems like something that might be worth adding unit tests for. It might make sense to pull some of these conversion utilities out into their own module and adding a unit test suite for that module.

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx [hunk 22]
```diff
@@ -434,134 +604,129 @@ const FormattedTextEditorView = forwardRef<
...
-						const retainedStr = content.slice(utf16Pos, utf16Pos + op.retain);
-						const cpCount = codepointCount(retainedStr);
-
-						if (op.attributes) {
+			if (source !== "user") return;
+			runOnce(isUpdating, () => {
+				// Wrap all tree mutations in a transaction so they undo/redo as one atomic unit.
+				// If the node is not part of a branch (e.g. unhydrated), apply edits directly.
+				const branch = TreeAlpha.branch(root);
+				const applyDelta = (): void => {
```
Comment (Josmithr) [inline packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Can we hoist this logic out into a helper function? Would be good to unit test the Quill deltas -> ArrayNode deltas conversion

**Review by Josmithr — COMMENTED** [review]
Left some comments. Mostly about code factoring and unit tests.

Overall, looking good. But there is a lot of complexity (in particular in the array-node delta <-> quill delta conversions) that I think we should ensure is factored out / tested.

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 11 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts]: renamed str -> value

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 12 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts:56]: added example blocks to both functions

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 13 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts]: Updated with comment

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 15 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts:1]: Added `codePointUtils.spec.ts` with test suite

File: packages/framework/quill-react/src/shared/runOnce.ts (hunk 17 above)
Comment (daesunp) [reply packages/framework/quill-react/src/shared/runOnce.ts]: Updated name to `runGuarded` and updated docs to be more clear

File: packages/framework/quill-react/src/plain/quillView.tsx (hunk 18 above)
Comment (daesunp) [reply packages/framework/quill-react/src/plain/quillView.tsx]: The early return on the `if(isUpdatingRef.current) return is inside the `runGuarded` itself, so the body is skipped when an outgoing direction sync is in progress. But I think the name was a bit misleading to I renamed to `runGuarded`

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx (hunk 20 above)
Comment (daesunp) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Updated to use `AttributeMap`

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx (hunk 21 above)
Comment (daesunp) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx:128]: added test suite for `applyQuillDeltaToTree` in `applyQuillDeltaToTree.tests.ts`.

I noticed that the conversion utilities are only used in `quillFormattedView.tsx` though, should we still move these into their own modules in case we want to reuse them for the future? Happy to make this change, just wanted to confirm!

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx (hunk 22 above)
Comment (daesunp) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Moved to a `applyQuillDeltaToTree` function and also has test suite now

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx (hunk 19 above)
Comment (daesunp) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: moved to `sizeToQuillAttribute` function

File: packages/dds/tree/src/test/text/textDomain.spec.ts [hunk 23]
```diff
@@ -80,6 +80,132 @@ describe("textDomain", () => {
 		});
 	});
 
+	describeHydration("onCharactersChanged", (_init, hydrated) => {
```
Comment (Josmithr) [inline packages/dds/tree/src/test/text/textDomain.spec.ts:83]: Nit: it would probably be worth adding tests for insertion / removal of empty ranges. I'm assuming we get no change notification in those cases, but it would be good to capture that explicitly.

File: packages/dds/tree/src/test/text/codePointUtils.spec.ts [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+
+		it("starts measuring from `start`", () => {
+			// "a😀b": skip 'a' (1 unit), measure 1 code point ('😀') = 2 units
+			assert.equal(utf16LengthForCodePoints("a😀b", 1, 1), 2);
+			// skip 'a' + '😀' (3 units), measure 1 code point ('b') = 1 unit
+			assert.equal(utf16LengthForCodePoints("a😀b", 3, 1), 1);
+		});
+
+		it("throws UsageError when start or count is out of range", () => {
+			assert.throws(() => utf16LengthForCodePoints("abc", -1, 1), UsageError); // negative start
```
Comment (Josmithr) [inline packages/dds/tree/src/test/text/codePointUtils.spec.ts]: We actually have a `validateUsageError` utility function you can use here to ensure the error has the right message, etc. I would recommend using that for these failure mode test cases.

File: packages/dds/tree/src/test/text/textDomain.spec.ts (hunk 23 above)
Comment (Josmithr) [reply packages/dds/tree/src/test/text/textDomain.spec.ts:83]: (Same for the formatted text tests)

File: packages/dds/tree/src/text/textDomain.ts [hunk 25]
```diff
@@ -186,6 +200,62 @@ class StringArray extends sf.array("StringArray", SchemaFactory.string) {
 	}
 }
 
+/**
+ * Processes an array-node delta into a {@link TextAsTree.TextOp}[] and calls `callback`.
+ * Shared by both the plain `onCharactersChanged` (from `nodeChanged`) and formatted `onContentChanged`
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomain.ts:206]: Nit
```suggestion
 * @remarks
 * Shared by both the plain `onCharactersChanged` (from `nodeChanged`) and formatted `onContentChanged`
```

File: packages/dds/tree/src/text/textDomain.ts [hunk 26]
```diff
@@ -186,6 +200,62 @@ class StringArray extends sf.array("StringArray", SchemaFactory.string) {
...
+ * retain ops include an explicit `formattingChanged: boolean`. Otherwise `formattingChanged` is omitted.
+ * @param getChar - Returns the character string at the given array index in the **post-edit** tree.
+ * Only invoked for insert ops, where it must read the inserted character at the given index of the tree
+ * after the edit has been applied. Passing an accessor that reads pre-edit content will silently produce wrong text.
+ * Return `undefined` if the tree is out of sync with the delta; this triggers a full-reread fallback.
+ * @param callback - The user-supplied callback to invoke with the translated ops.
+ */
+export function processCharactersChangedDelta(
+	delta: readonly (ArrayNodeDeltaOp | ArrayNodeTreeChangedDeltaOp)[] | undefined,
+	getChar: (index: number) => string | undefined,
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomain.ts]: Nit: we generally prefer to avoid abbreviations when possible. Can we call this `getCharacter`?

File: packages/dds/tree/src/text/textDomain.ts [hunk 27]
```diff
@@ -186,6 +200,62 @@ class StringArray extends sf.array("StringArray", SchemaFactory.string) {
...
+	callback: (ops: readonly TextAsTree.TextOp[] | undefined) => void,
+): void {
+	if (delta === undefined) {
+		callback(undefined);
+		return;
+	}
+	let readPos = 0;
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomain.ts]: Nit: `readPosition`?

File: packages/dds/tree/src/text/textDomain.ts [hunk 28]
```diff
@@ -249,6 +319,69 @@ class StringArray extends sf.array("StringArray", SchemaFactory.string) {
...
+		/**
+		 * Discriminator identifying this op as a retain.
+		 */
+		readonly type: "retain";
+		/**
+		 * The number of Unicode code points to retain.
+		 */
+		readonly count: number;
+		/**
+		 * Present only on retain ops delivered by {@link @fluidframework/tree#FormattedTextAsTree.Members.onContentChanged};
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomain.ts:340]: Nit: we should reformat this to state what it _is_ first, then add any deeper notes in a `@remarks` block.

File: packages/dds/tree/src/text/textDomain.ts [hunk 29]
```diff
@@ -263,6 +396,7 @@ export namespace TextAsTree {
 
 	/**
 	 * Interface for a text node.
+	 * @sealed
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomain.ts]: Nit: we usually put `@sealed` near the bottom by the release tag.

File: packages/dds/tree/src/text/textDomain.ts [hunk 30]
```diff
@@ -325,6 +459,22 @@ export namespace TextAsTree {
...
+		 * or `undefined` when a delta could not be computed (e.g. during a schema upgrade).
+		 * @returns A cleanup function that unsubscribes the callback when called.
+		 * @remarks
+		 * Only fires on shallow changes — inserts and removes.
+		 * It does not fire on deep changes such as formatting property updates on existing characters.
+		 * For formatted text, use {@link @fluidframework/tree#FormattedTextAsTree.Members.onContentChanged} to also receive deep changes.
+		 *
+		 * All counts in the delivered ops are in Unicode code points, not UTF-16 code units.
+		 * For characters outside the Basic Multilingual Plane (e.g. emoji), one code point
+		 * corresponds to two UTF-16 code units — convert before using the counts as string indices.
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomain.ts:478]: Great docs! Thanks!!

File: packages/dds/tree/src/text/textDomainFormatted.ts [hunk 31]
```diff
@@ -169,6 +173,11 @@ class TextNode
 			// If this node does not have a corresponding branch, then it is unhydrated.
 			// I.e., it is not part of a collaborative session yet.
 			// Therefore, we don't need to run the edits as a transaction.
+			// Note: for unhydrated nodes each atom edit fires a separate `treeChanged` event,
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomainFormatted.ts:176]: Can we mitigate this somehow? Could we leverage `withBufferedEvents` for this?

File: packages/dds/tree/src/text/textDomainFormatted.ts [hunk 32]
```diff
@@ -177,6 +186,34 @@ class TextNode
 			});
 		}
 	}
+	/**
+	 * Returns the character string at the given atom index, or `undefined` if out of bounds.
+	 * Line atoms expand to `"\n"`; text atoms return their underlying code point(s).
+	 */
+	private atomChar(i: number): string | undefined {
```
Comment (Josmithr) [inline packages/dds/tree/src/text/textDomainFormatted.ts]: Nit: we generally recommend naming functions/methods using verbs. Also recommended to avoid abbreviations where possible. How about `getAtomCharacterAt` (or something)?

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx [hunk 33]
```diff
@@ -285,36 +308,324 @@ function quillAttributesToPartial(
...
-				: `${format.size}px`;
+		attributes.size = sizeToQuillAttribute(format.size);
 	}
 	if (format.font !== defaultFont) attributes.font = format.font;
 	return attributes;
 }
 
+/**
+ * Convert a CharacterFormat to Quill attributes that fully describe the formatting state.
+ * Unlike {@link formatToQuillAttributes}, this includes `null` for default-valued properties
```
Comment (Josmithr) [inline packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Nit:
```suggestion
 * @remarks
 * Unlike {@link formatToQuillAttributes}, this includes `null` for default-valued properties
```

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx (hunk 21 above)
Comment (Josmithr) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx:128]: I still think it would probably make sense, yeah. These are complex enough that they deserve unit test coverage, in my opinion. And then we can rename the test module after the module we create for this stuff.

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 11 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts]: Updated with suggestion

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 13 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts]: Updated with explanation comment

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 14 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts]: Updated to throw usage errors

File: packages/dds/tree/src/text/codePointUtils.ts (hunk 15 above)
Comment (daesunp) [reply packages/dds/tree/src/text/codePointUtils.ts:1]: Added test suite

File: packages/dds/tree/src/test/text/codePointUtils.spec.ts (hunk 24 above)
Comment (daesunp) [reply packages/dds/tree/src/test/text/codePointUtils.spec.ts]: Updated to use `validateUsageError`

File: packages/dds/tree/src/text/textDomain.ts (hunk 25 above)
Comment (daesunp) [reply packages/dds/tree/src/text/textDomain.ts:206]: Updated with suggestion

File: packages/dds/tree/src/text/textDomain.ts (hunk 27 above)
Comment (daesunp) [reply packages/dds/tree/src/text/textDomain.ts]: Updated with suggestion

File: packages/dds/tree/src/text/textDomain.ts (hunk 26 above)
Comment (daesunp) [reply packages/dds/tree/src/text/textDomain.ts]: Updated with suggestion

File: packages/dds/tree/src/text/textDomain.ts (hunk 28 above)
Comment (daesunp) [reply packages/dds/tree/src/text/textDomain.ts:340]: Updated with suggestion

File: packages/dds/tree/src/text/textDomain.ts (hunk 29 above)
Comment (daesunp) [reply packages/dds/tree/src/text/textDomain.ts]: Updated with suggestion

File: packages/dds/tree/src/text/textDomainFormatted.ts (hunk 32 above)
Comment (daesunp) [reply packages/dds/tree/src/text/textDomainFormatted.ts]: Updated with suggestion

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx (hunk 33 above)
Comment (daesunp) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Updated with suggestion

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx (hunk 21 above)
Comment (daesunp) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx:128]: Moved to quillAttributeUtils.ts and added unit tests

File: packages/dds/tree/src/text/textDomainFormatted.ts (hunk 31 above)
Comment (daesunp) [reply packages/dds/tree/src/text/textDomainFormatted.ts:176]: Unfortunately when using `withBufferedEvents` and we touch two fields, we invalidate and return undefined, so we can't use it as a workaround.

Updated the docs mentioning this

Comment (Josmithr) [reply packages/dds/tree/src/text/textDomainFormatted.ts:176]: We should file a bug for this and probably reference it here. I think any time the hydrated vs unhydrated behavior differs, that constitutes a bug that we need to (eventually) fix.

Comment (Josmithr) [reply packages/dds/tree/src/text/textDomainFormatted.ts:176]: Do we have test coverage for this? If not, we should add it and similarly document the bug and what the intended behavior should be.

File: packages/framework/quill-react/src/formatted/quillFormattedView.tsx [hunk 34]
```diff
@@ -244,77 +105,272 @@ function createLineAtom(
...
-	if (format.italic) attributes.italic = true;
-	if (format.underline) attributes.underline = true;
-	if (format.size !== defaultSize) {
-		// Convert pixel value back to Quill size name if possible
-		attributes.size =
-			format.size in sizeReverse
-				? sizeReverse[format.size as keyof typeof sizeReverse]
-				: `${format.size}px`;
+export function applyQuillDeltaToTree(root: FormattedTextAsTree.Tree, delta: Delta): void {
+	const branch = TreeAlpha.branch(root);
```
Comment (brrichards) [inline packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Should this be using `TreeAlpha.context().isBranch()` with a transaction label here? The plain text view passes `label: editLabelRef.current ` through `context.runTransaction()`, but this function uses TreeAlpha.branch() and drops the label. I think without the label, formatted editor edits won't be reachable by the scoped undo button since they'd be unlabeled and only show up in the global undo stack.

Comment (daesunp) [reply packages/framework/quill-react/src/formatted/quillFormattedView.tsx]: Added optional label parameter to `applyQuillDeltaToTree` and routed through `TreeAlpha.context().runTransaction(applyDelta, { label })`

**Review by brrichards — APPROVED** [review]
LGTM!
